File: src/main/java/community/whatever/onembackendjava/repository/ShortenedUrlRepository.java
package community.whatever.onembackendjava.repository;

import java.util.Optional;

import community.whatever.onembackendjava.entity.ShortenedUrlEntity;

public interface ShortenedUrlRepository {
	/**
	 * <p> 단축된 Url로 조회</p>
	 *
	 * @param shortenedUrl 단축된 Url
	 * @return 단축 Url entity
	 */
	Optional<ShortenedUrlEntity> findByShortenedUrl(String shortenedUrl);

	/**
	 * <p>원본 Url과 단축 Url을 저장</p>
	 *
	 * @param shortenedUrlEntity 저장할 단축 Url 객체
	 * @return 저장된 단축 Url entity
	 */
	ShortenedUrlEntity save(ShortenedUrlEntity shortenedUrlEntity);

}

File: src/main/kotlin/community/whatever/onembackendkotlin/domain/ShortenedUrlRepository.kt
package community.whatever.onembackendkotlin.domain

import java.time.LocalDateTime
import java.util.Optional

interface ShortenedUrlRepository {
    /**
     * 아이디를 통해 원본 URL을 찾는다.
     *
     * @param id ShortenedUrl의 id이다.
     * @return 존재하면 ShortenedUrl을 반환하고 존재하지 않으면 null을 반환한다.
     */
    fun findById(id: String): Optional<ShortenedUrl>

    fun findByIdAndDeletedIsFalse(id: String): Optional<ShortenedUrl>

    /**
     * ShortenedUrl을 저장한다.
     *
     * @param shortenedUrl 저장할 ShortenedUrl이다.
     * @return 저장된 ShortenedUrl을 반환한다.
     */
    fun save(shortenedUrl: ShortenedUrl): ShortenedUrl

    /**
     * 원본 URL을 통해 ShortenedUrl을 찾는다.
     *
     * @param originUrl ShortenedUrl의 원본 URL이다.
     * @return 존재하면 ShortenedUrl을 반환하고 존재하지 않으면 null을 반환한다.
     */
    fun existsByOriginUrl(originUrl: String): Boolean

    /**
     * 원본 URL을 통해 ShortenedUrl을 찾는다.
     *
     * @param originUrl ShortenedUrl의 원본 URL이다.
     * @return 존재하면 ShortenedUrl을 반환하고 존재하지 않으면 null을 반환한다.
     */
    fun findByOriginUrl(originUrl: String): Optional<ShortenedUrl>

    /**
     * 모든 ShortenedUrl을 삭제한다.
     */
    fun deleteAll()

    /**
     * 만료일이 주어진 시간 이전인 ShortenedUrl을 모두 삭제한다.
     *
     * @param baseTime 기준 시간이다.
     */
    fun deleteAllByExpiredAtBefore(baseTime: LocalDateTime)
}
